scripts: zephyr_module: Add CPE support

tgagneret-embedded opened the pull request:
This allows the possibility of:

 - Adding CPE for modules when generating the SBOM (SPDX)
 - Monitoring CVE by adding a CPE related to modules (e.g a module is a forked of an other repository)

`cpe` and `version` fields are extracted from module.yml and included in generated kernel meta file (making it available for SBOM generation).

This PR is related to issue #53479 and follows PR #66182 that aims to improve the SBOM output and better monitoring of vulnerabilities.

tejlmand commented on doc/develop/modules.rst:
```diff
@@ -569,6 +569,36 @@ Build files located in a ``MODULE_EXT_ROOT`` can be described as:
+
+.. code-block:: yaml
+
+   vulnerabilities:
```
not sure `vulnerabilities` is the best key at this level.
It can give the impression that it will contain information regarding actual vulnerabilities, which is not the case.

You already propose a `cpe` entry, so why can we not provide the `remote` as a sub-field under there ?

tgagneret-embedded replied:
An evolution to `vulnerabilities` would be to add the field `vex` which would point to a vex file allowing to filter CVE.
Do you think it would fit in `cpe` field or should we find an other name (not sure what it would be though) ?

tejlmand replied:
thanks for sharing some thoughts for future direction of this.

With a `vex` (or vdr) field, then vulnerabilities starts to make more sense as name.
I don't have a better suggestion for another name.

kartben replied:
how about simply`security:`? `security_info:`? 
I do agree with @tejlmand that "vulnerabilities" makes it sound like it lists vulnerabilities or something

tejlmand replied:
I like security_info more.
Broad enough to cover the vex info, while still specific enough to be useful for readers to understand the context.

tejlmand commented on doc/develop/modules.rst:
```diff
@@ -569,6 +569,36 @@ Build files located in a ``MODULE_EXT_ROOT`` can be described as:
 This allows control of the build inclusion to be described externally to the
 Zephyr module.
 
+SBOM integration
```
Should we turn this around.

Instead of having `SBOM integration`, then describe what content is (example: `Vulnerability monitoring`), and then instead describe that this information will be included in the sbom.

The vulnerability monitoring / information can be of interest, even without an SBOM.

(Just a thought, not a request for a change)

tejlmand commented on doc/develop/modules.rst:
```diff
@@ -569,6 +569,36 @@ Build files located in a ``MODULE_EXT_ROOT`` can be described as:
+The module description file :file:`zephyr/module.yml` can be used to improve the
+generated SBOM and vulnerability monitoring.
+
+The version and the CPE associated to the module can be provided using
```
when using abbreviations like: `CPE` and `CVE` (and soon `VEX`), we should write them in full once, so that even users unfamiliar with the terms know what they mean.

tejlmand commented on doc/develop/modules.rst:
```diff
@@ -569,6 +569,36 @@ Build files located in a ``MODULE_EXT_ROOT`` can be described as:
+
+.. code-block:: yaml
+
+   version: <module-version>
```
Let's move the `version` out of `SBOM integration`.

`version` has value on its own, and should be considered first class citizen in the module yaml.
Just like `name` is independent: https://docs.zephyrproject.org/latest/develop/modules.html#module-name

tejlmand reviewed (changes requested):
Thanks for this addition.

Notice some minor parts to adjust before +1, namely `version` and first time use of abbreviations.

tgagneret-embedded commented:
@tejlmand I've taken into account your review. I also added the integration of CPE and version in the SBOM since documentation talk about the SBOM.

I wonder if we should have some testing on the output SBOM ? For example we could generate it for each project built in the pipeline and then run [pyspdxtools](https://github.com/spdx/tools-python) to validate them. That would be a start. What do you think ?

I'm not familiar with zephyr pipeline, could you point me to the correct files if you think testing is a requirement for this.

Thanks

tgagneret-embedded commented:
I wonder if I should add `remote_purl` to the `vulnerabilities` field ? That would make a lot of sense for project who only have github advisory database support (MCUBoot for example).

Maybe I could rename `remote_cpe` to `external_references` and the list could be either `purl` or `cpe`. Parsing would detect automatically which one it is.

I would keep `cpe` field but I don't know if a new `purl` field is necessary right now, since it would only make sense if someone mirrors modules of the project. Do you think it should be added as well ?

What do you think @tejlmand ?

tgagneret-embedded commented:
I've added a new commit with the modification from my previous comment. If it's okay with you, I can squash them at the end.

tgagneret-embedded committed:
scripts: zephyr_module: Add URL, version to SPDX

Improve the SPDX with the current values:
 - URL: extracted from `git remote`. If more than one remote, URL is not
 set.
 - Version: extracted from `git rev-parse` (commit id).
 - PURL and CPE for Zephyr: generated from URL and version.

For zephyr, the tag is extracted, if present, and replace the commit id for
the version field.
Since official modules does not have tags, tags are not yet extracted for
modules.

To track vulnerabilities from modules dependencies, a new SBOM,
`modules-deps.spdx` was created. It contains the `external-references`
provided by the modules. It allows to easily track vulnerabilities from
these external dependencies.

Signed-off-by: Thomas Gagneret <thomas.gagneret@hexploy.com>

tejlmand commented on doc/develop/modules.rst:
```diff
@@ -493,6 +493,18 @@ Here is an example for the Zephyr module ``foo``:
+Module version
+==============
+
+Each Zephyr module should have a version number to improve tracking.
```
`should` seems a bit strong in this case, especially as it would indicate that modules are having some form of release cycles. This may be true for some modules, but other modules revision might be considered to be identified (thus indirectly follow the Zephyr version) through the SHA in the manifest, which again is identified by the Zephyr revision / version.

```suggestion
Each Zephyr module may have a version number to improve tracking.
```

tejlmand commented on doc/develop/modules.rst:
```diff
@@ -569,6 +581,34 @@ Build files located in a ``MODULE_EXT_ROOT`` can be described as:
+
+The module description file :file:`zephyr/module.yml` can be used for vulnerability monitoring
+and improve the generated SBOM output.
+
```
Very good.

Maybe at this location specify the `version:` field should be used.
For example:
```suggestion

When a module monitors vulnerabilities then the module should also define the `version` field, :ref:`module_version`.
```

tgagneret-embedded replied:
Your comment makes me think this `version` field creates confusion. There was a misunderstanding from my part in the original issue.

The goal was to provide a version for the module and help complete the SBOM for the module (and not searching vulnerabilities from the forked repository for example). The `external-references` field must be used for that (and CPE and PURL specifications allow to include the version)

I think we should remove the `version` field from this PR and I will clarify this for a next PR. Are you okay with this ?

tgagneret-embedded replied:
It seems that there are some clarification (raised by your comment) to be done on this PR with the security group. In the meantime, I converted it into draft and I'll update it when I have all the information I need.

I'm sorry for this.

tejlmand replied:
if you believe it clarifies the usage and intentions of the field, then it's fine.

The clearer the intention of a PR, the better.

tejlmand reviewed (dismissed):
Looks good to me.

So minor doc suggestions but nothing blocking.

ceolin commented:
I tried 

```
version: "3.5"
vulnerabilities:
  external-references: 
    - "cpe:2.3:a:arm:mbed_tls:*:*:*:*:*:*:*:*"
    - "pkg:https://github.com/Mbed-TLS/mbedtls/"
build:
  cmake-ext: True
  kconfig-ext: True 
```
in mbedTLS module and I couldn't see the cpe in the output, if I do: 

```
version: "3.5"
cpe: "cpe:2.3:a:arm:mbed_tls:*:*:*:*:*:*:*:*"
build:
  cmake-ext: True
  kconfig-ext: True 
```

I can see the cpe, but my understanding is that we should see cpe for downstream and upstream packages right ? Though we will probably be using (at least, in beginning) the upstream CPE.

tgagneret-embedded commented:
@ceolin Were you able to test it ? Does it look good to you ?

Thanks

ceolin commented:
> @ceolin Were you able to test it ? Does it look good to you ?
> 
> Thanks

Hi !  It is looking correct the output to me. pyspdxtools parse properly the output, but cve-bin-tool was not able to find vulnerabilities in a packages properly, it seems to ignore external references. I tried one cpe that I know has vulnerabilities associated with ( cpe:2.3:a:arm:mbed_tls:1.3.3:*:*:*:*:*:*:* ). modules-deps.spdx has:
```
##### Package: mbedtls-deps

PackageName: mbedtls-deps
SPDXID: SPDXRef-mbedtls-deps
PackageLicenseConcluded: NOASSERTION
PackageLicenseDeclared: NOASSERTION
PackageCopyrightText: NOASSERTION
PackageDownloadLocation: NOASSERTION
ExternalRef: SECURITY cpe23Type cpe:2.3:a:arm:mbed_tls:1.3.3:*:*:*:*:*:*:*
ExternalRef: PACKAGE_MANAGER purl pkg:https://github.com/Mbed-TLS/mbedtls/
FilesAnalyzed: false
PackageComment: Utility target; no files
```

If I try:
`cve-bin-tool --sbom spdx --sbom-file build/spdx/modules-deps.spdx `
![image](https://github.com/zephyrproject-rtos/zephyr/assets/882919/50d8be83-4ef6-479c-b18a-013a9f02a1ba)

Then, if I changed modules-deps.spdx to
```
PackageName: mbed_tls
PackageVersion: 1.3.3
SPDXID: SPDXRef-mbedtls-deps
PackageLicenseConcluded: NOASSERTION
PackageLicenseDeclared: NOASSERTION
PackageCopyrightText: NOASSERTION
PackageDownloadLocation: NOASSERTIONy
ExternalRef: SECURITY cpe23Type cpe:2.3:a:arm:mbed_tls:1.3.3:*:*:*:*:*:*:*
ExternalRef: PACKAGE_MANAGER purl pkg:https://github.com/Mbed-TLS/mbedtls/
FilesAnalyzed: false
PackageComment: Utility target; no files
```
![image](https://github.com/zephyrproject-rtos/zephyr/assets/882919/e47a7271-e5b2-444a-82a3-ad9fe9e631f5)


It seems to me that cve-bin-tool should be get values from ExternalRef and the spdx here is correct, but  ...

That's why I mentioned to have packagename and version but it may just be a tool problem.

It is not part of this pr, but we need to relate the package with the source at some point, as you can see the generated output says that no files were analyzed which is not correct, but it is not related with these changes :)

tgagneret-embedded commented:
Hi @ceolin,

cve-bin-tool does not parse everything in external refs. For now, only PURL containing a version are taken into account, but CPE are not.
I have created a PR on cve-bin-tool to fix this CPE issue (https://github.com/zephyrproject-rtos/zephyr/pull/66495).

I guess if it's okay with you, @tejlmand @nordicjm, PR is finalized and ready to be reviewed.

Thanks

ceolin reviewed (dismissed):
@tgagneret-embedded Thanks ! This is really helpful.

kartben commented:
@tgagneret-embedded just curious, have you started looking at integrating SPDX 3?

tgagneret-embedded commented:
Can I have some update on this PR @tejlmand  ?

Thank you :)

kartben commented on doc/develop/modules.rst:
```diff
@@ -569,6 +569,36 @@ Build files located in a ``MODULE_EXT_ROOT`` can be described as:
+.. code-block:: yaml
+
+   vulnerabilities:
+     external-references:
```
I *think* I'd be more specific and make this two lists: `cpe-identifiers:` and `purl-identifiers:` as `external-references:` sounds really vague (unless you do want to allow people to potentially add whatever types of external-references they want, but at least now it looks like an external-identifier entry that matches neither of the two regexps gets dropped)

kartben commented on doc/develop/modules.rst:
```diff
@@ -569,6 +569,36 @@ Build files located in a ``MODULE_EXT_ROOT`` can be described as:
+If your module needs to track vulnerabilities using an external reference
+(e.g your module is forked from an other repository), you can use the ``vulnerabilities`` section.
+It contains the field ``external-references`` that contains a list of references that needs to
+be monitored for your module. The supported format are:
```
```suggestion
be monitored for your module. The supported formats are:
```

kartben commented on doc/develop/modules.rst:
```diff
@@ -569,6 +569,36 @@ Build files located in a ``MODULE_EXT_ROOT`` can be described as:
+- CPE (Common Platform Enumeration)
+- PURL (Package URL)
+
+.. code-block:: yaml
```
it might be my own preference but I would also add an actual example of how things would look like with real values, ex. by using mbed TLS, or whatever, as an example. Also, as a way to make sure people instantly realize where the new section belongs, i would have the YAML snippet show more than just the relevant section, ex. include name and build entries

```
name: lvgl
build:
  ...
vulnerabilities:
   external-references:
     - ...
```

swinslow commented on scripts/west_commands/zspdx/walker.py:
```diff
@@ -278,6 +340,42 @@ def setupSDKDocument(self):
+        cfgModuleExtRef.docRefID = "DocumentRef-modules-deps"
+        self.docModulesExtRefs = Document(cfgModuleExtRef)
+
+        for module in modules:
```
SPDX does not _require_ listing the entire set of modules. SPDX is essentially a language that lets you describe a set of software packages to a greater or lesser degree of specificity.

@kartben I'm happy to share some thoughts about how to potentially use these sorts of SPDX documents to "trace" the impact of a vulnerability, but agree with @tgagneret-embedded that it probably makes more sense in a separate issue or PR thread.

swinslow commented on scripts/west_commands/zspdx/sbom.py:
```diff
@@ -121,4 +121,10 @@ def makeSPDX(cfg):
+    # write modules document
+    writeSPDX(os.path.join(cfg.spdxDir, "modules-deps.spdx"), w.docModulesExtRefs)
+    if not retval:
+        log.err("SPDX writer failed for build document; bailing")
```
Minor nit, I'd probably change "build document" to "modules dependencies document" or something like that, for the error message here.

swinslow commented:
@tgagneret-embedded Thanks for this! Although I'm not as familiar with Zephyr modules (or the current SPDX generator code) anymore and I haven't had a chance to test this out, at a high level the changes in this PR make sense and look like a good addition to me.

kartben commented on doc/develop/modules.rst:
```diff
@@ -569,6 +569,45 @@ Build files located in a ``MODULE_EXT_ROOT`` can be described as:
+
+.. code-block:: yaml
+
+   security:
+     external-references:
+       - cpe:2.3:a:arm:mbed_tls:3.5.2:*:*:*:*:*:*:*
+       - pkg:github/Mbed-TLS/mbedtls@V3.5.2
```
This actually poses a good question regarding what CPE and PURLs should be used for the Zephyr forks of the upstream project. 
Will it be expected that e.g. the https://github.com/zephyrproject-rtos/mbedtls module shall reference the CPE/PURL of the most recently released version of the upstream mbedtls project?

BTW I think it would be great if the sample code block would show a bit more of the "surroundings" so that folks have an immediate clue regarding the nesting level of the `security` key without having to look at the schema, i.e

```
name: mbedtls
build:
  ...
security:
   external-references:
       - cpe:2.3:a:arm:mbed_tls:3.5.2:*:*:*:*:*:*:*
       - pkg:github/Mbed-TLS/mbedtls@V3.5.2
```

tgagneret-embedded replied:
The Zephyr forks must used the CPE/PURL with the version it forked for the current version. For example, for mbedTLS, the last commit merge the upstream 3.5.2 into the zephyr fork (https://github.com/zephyrproject-rtos/mbedtls/commit/6ec4abdcda78dfc47315af568f93e5ad4398dea0) , so `module.yml` must contain the example I described.

For the example, I did not want to add keys that are not related to my modification (it would probably not be updated if a changes occurs in `build` for example). Let me know what I should do.

kartben replied:
OK, I now understand why you wouldn't list other keys in case the schema changes, still wonder if it would make sense to at least mention something like " ... you can use the top-level ``security`` section" or something, just to help people know quicker where they need to put it

Re: CPE/PURL, part of my question was also (but really just a question for my own education and not something that should get in the way of this PR in any way) how to deal with the Zephyr module forks where it's more than just mirroring the upstream module but also including our own patches (which can come with their own vulnerabilities) on top of it. Should the "external references" list both the CPE/PURL of the upstream module + a CPE/PURL corresponding to the Zephyr fork?

tgagneret-embedded replied:
You are totally right. This was part of the initial discussion (#53479). At first I wanted zephyr modules maintainers to have the responsibility of the CVE for their project. We finally decided, for now, it wouldn't be the case. 
For now, CVE in zephyr modules are not reported (no CPE or via Github Advisiory Database). This is something I would like to change. 
This PR is part of multiple PR (I hope) that could lead to what you propose.

The future steps I would like to have:

 - add VEX support in `module.yml` (paths to VEX)
 - automate vulnerabilities detection from upstream (in zephyr pipeline)
 - add a CPE for each zephyr modules

But this is a discussion to have with the security team / @ceolin and see what's possible or not. Let's take this step by step :)

I hope it answers your question.

kartben replied:
> I hope it answers your question.

it very much does :) Thanks!

ceolin replied:
Yep, ideally Zephyr forks will have their on CPE, but adding it right now would be useless since these CPEs would no match any CVE database.  From a practical point of view, what we really need now is a way to monitor CVEs in the projects we have forked.

kartben commented on doc/develop/modules.rst:
```diff
@@ -569,6 +569,45 @@ Build files located in a ``MODULE_EXT_ROOT`` can be described as:
+The module description file :file:`zephyr/module.yml` can be used to improve vulnerability monitoring.
+
+If your module needs to track vulnerabilities using an external reference
+(e.g your module is forked from an other repository), you can use the ``security`` section.
```
```suggestion
(e.g your module is forked from another repository), you can use the ``security`` section.
```

kartben reviewed (dismissed):
Been playing with this over the last few days and it is really nice!
Just added a couple additional non-blocking comments

tgagneret-embedded commented:
@tejlmand Does it look good to you ?

tgagneret-embedded commented on scripts/west_commands/zspdx/walker.py:
```diff
@@ -216,39 +249,67 @@ def setupZephyrDocument(self, modules):
 
+        zephyr_url = zephyr.get("remote", "")
+        if zephyr_url:
+            cfgPackageZephyr.url = zephyr_url
```
I added the `generateDowloadUrl` function to follow SPDX specs. Only git is supported.

tgagneret-embedded commented:
@ceolin @kartben It seems the PR is stalled. I think every comments have been taken into account. It's been over a month and there is no new review or comment.

Can you help me with this ?

Thanks

kartben commented:
@tgagneret-embedded Thank you for your patience. It's "unfortunate" you just pushed a new commit as this dismissed the existing reviews and will only contribute to merge taking even longer, as this now needs another (hopefully much faster though) round of reviews. 

@tejlmand, please try to have a look at this PR again as it's been several months since it's been waiting for your approval as the assignee.

kartben commented on scripts/west_commands/zspdx/writer.py:
```diff
@@ -42,6 +51,18 @@ def writeFileSPDX(f, bf):
+def generateDowloadUrl(url, revision):
+    # Only git is supported
+    # walker.py only parse revision if it's from git repositiory
+    if len(revision) == 0 or url.startswith("git://"):
+        return url
+
+    url = url.replace("https://", "git+https://")
+    url = url.replace("http://", "git+http://")
+    url = "@".join([url, revision])
```
not sure this handle properly the case where url already starts with "git+http"? Shouldn't it?

tgagneret-embedded replied:
URL si extracted from meta file. The URL is the one for cloning the repository. I don't think it can have the url already starting with "git+".
This  function is not generic and takes into account that this script is run in zephyr: only git is supported, and url is parsed from ``git remote`` command (and maybe using ```west list``` values later)

So, do you think I should add a special case for the URL that starts with "git+" (even though  I don't think it can happen in our case) ?
Maybe you see a special case I did not see.

tejlmand reviewed (dismissed):
Much better. Still a couple of observations.

As for the last two commits, can you please squash those into the appropriate parent commits ?
When reviewing commit-by-commit, then it's not nice to suddenly see code that one comment suddenly being changed in later commit in same PR.

tgagneret-embedded commented on scripts/zephyr_module.py:
```diff
@@ -447,9 +430,91 @@ def git_revision(path):
+
+    revision, dirty = git_revision(path)
+    workspace_dirty |= dirty
+    remote, dirty = git_remote(path)
```
If you have multiple remotes in one module, how you make sure the current commit is from one remote or an other ?

One solution could be to use```git branch -a --contains <commit id>```, but I'm not sure it's reliable to determine if it's one commit is from a specific remote.
What happens if the commit is present in multiple remote ? Which one should I choose ?

So if I detect multiple remote, I mark the repository has "dirty" and no remote is selected.

How do you think we should handle this case (where remotes are added manually) ?

tejlmand replied:
> If you have multiple remotes in one module, how you make sure the current commit is from one remote or an other ?

You can't. A given commit SHA is unique, that's the nature of git.
But that doesn't make the remote "dirty".

In fact, you should not rely on the remote at all for this.
If you want to rely on something, then you should the remote from the `west` project, if the module is also a west project.

tgagneret-embedded replied:
> You can't.

I was thinking of a fork. So it's possible :)

tejlmand replied:
> > You can't.
>
> I was thinking of a fork. So it's possible :)

Not necessarily, and especially not with `github`.

If forking a repo using `github`'s fork feature, then you can actully fetch SHAs directly regardless of the fork to which it was pushed.

tgagneret-embedded commented on scripts/zephyr_module.py:
```diff
@@ -447,9 +430,91 @@ def git_revision(path):
+            remotes_name = stdout.rstrip().split('\n')
+
+        remote_url = None
+        if len(remotes_name) == 1:
```
> so http:// or https:// without git, doesn't seem to be sufficient.

The URL is transformed in ```writer.py``` because it's related to SPDX.
Just like you said that `-` should be replaced in ```writer.py``` and not in ```walker.py```.

> Note, a Zephyr module doesn't have to be a west project, also it doesn't even need to be a git repo.
So those cases must also be properly handled.

I agree, but before my PR, only git repository was supported. The only way to get the revision was through the ```git_revision``` function. So even if I agree with you, I would prefer to keep the current state of support and extend it later if possible.

I will have a look at how to get the same data returned by ```west list```, however I would need to know how I should handle the case where one or more remotes are added manually to a module.
It echoes a previous comment but:
What if a repository is checked out to a commit presents in a different remote than the one provided by ```west list```. What if the commit is in multiple remotes (and not necessarily in the one listed in ```west list```) ? Which remote should I choose ?

tejlmand replied:
> What if a repository is checked out to a commit presents in a different remote than the one provided by west list.

That is why the build output meta appends `-off` to the commit when the sha used is different from the sha defined in the manifest.
https://docs.zephyrproject.org/latest/kconfig.html#CONFIG_BUILD_OUTPUT_META

Even with your proposal of relying on a single remote will provide wrong information in case a user does a local commit, because in such a case the sha is not available at the remote.

Also users can do a `git fetch <url>` followed by a `git checkout FETCH_HEAD`.
In such a case there may be a single remote, but a commit from different url is checked out locally.
Such workflow is very common for testing PRs or branches.

> What if the commit is in multiple remotes (and not necessarily in the one listed in west list) ? Which remote should I choose ?

When `west` is used, then use the `west remote` if the sha is identical to the sha defined by the current manifest.
Because in that case we know the sha can be reached by that remote.
If it's not identical, then consider the sha `-off`. No reason to re-invent what `CONFIG_BUILD_OUTPUT_META` already provides. (Have you looked at this feature at all ?)

tgagneret-embedded replied:
I will use the `-off`. Thanks.
So, if I understand correctly, even though there is multiple remotes, I will use the remote returned by `west list`.

tgagneret-embedded replied:
@tejlmand I don't think there is any remote provided by west for the manifest repository.

If running `west init -m https://github.com/zephyrproject-rtos/zephyr.git`, `west list` does not provide any remote for the manifest. I didn't find anything in west code to have this information (command or function), you can only extract the manifest directory. The only solution I have is to run `git remote` in the manifest directory to get it.

So if I have two remotes (for example, zephyr remotes and the forked zephyr to add my feature), I don't have any clue on which remotes I should use.

What should I do in this case ? No remote ? The first one ?

Thanks.

kartben commented:
re: @tejlmand's comment below (which I can't reply to as it's "outdated)

> I like security_info more.
> Broad enough to cover the vex info, while still specific enough to be useful for readers to understand the context.

It's now `security`, actually, but it looks like the commit that introduced it hasn't been squashed properly so the initial name also shows up in the history.

tgagneret-embedded commented:
>@tejlmand I don't think there is any remote provided by west for the manifest repository.
>
>If running west init -m https://github.com/zephyrproject-rtos/zephyr.git, west list does not provide any remote for the >manifest. I didn't find anything in west code to have this information (command or function), you can only extract the >manifest directory. The only solution I have is to run git remote in the manifest directory to get it.
>
>So if I have two remotes (for example, zephyr remotes and the forked zephyr to add my feature), I don't have any clue on >which remotes I should use.
>
>What should I do in this case ? No remote ? The first one ?
>
>Thanks.

On top of this question, should we use `projects` instead of `modules` in the following code (for the content of `zephyr.spdx`) ?

https://github.com/zephyrproject-rtos/zephyr/blob/b306e80e00d71196913228cc96cecb717d5357a1/scripts/west_commands/zspdx/walker.py#L292C44-L292C60

Maybe `zephyr.spdx` should contains only the modules. Thus should we make a new spdx for the projects ?

For example, when we use an external repository for the manifest, this manifest won't appear in any SPDX/SBOM. So where should it appears ? In a broader sense, if someone add new projects in manifest, it's not listed in any SPDX/SBOM.

By the way it seems that all modules are listed as a project. I don't know if it's possible to have a module not listed as a project ?

I think the creation of a new SPDX `projects.spdx` makes sense, however I'm not sure `zephyr.spdx` would still be useful.
Does anyone have a suggestion about this ?


Thanks

tgagneret-embedded committed:
scripts: zephyr_module: Move SPDX name normalization to writer.py

Since `writer.py` is the one writting the SPDX file, it should normalize
the name field and not `walker.py` which generates the SBOM components.

Signed-off-by: Thomas Gagneret <thomas.gagneret@hexploy.com>

tgagneret-embedded commented:
@tejlmand 

 - I now support the "-off" flag when revision is not the same as the one present in the manifest.
 - I use the URL from the manifest if I have no remote, more than one or a different remote (and add the "-off" flag to the revision)
 - Modules now support the "-off" (Since modules provided by west does not provide the URL, I check if the module is not in the project list to get the required information)
 - I fixed the "normalization of the SPDX field" and move it to `writer.py` (instead of `walker.py`). I had to normalize the relationships too, so it's not very pretty in my opinion. It's in a separate commit (https://github.com/zephyrproject-rtos/zephyr/pull/66495/commits/b1452527c55409215c092827f81c253bdfc7b03b) and I will drop it if you are not satisfied.

tgagneret-embedded commented:
@tejlmand Can you tell me if there anything missing from my part for you to review this PR ? If so, let me know, I would be happy to fix it. This PR is opened for a long time now 😄.

Thanks

kartben commented:
ping @tejlmand

kartben commented:
@tgagneret-embedded although it's getting late in the release cycle for doing this, would it make sense to submit updates to some of the manifests of the modules where this new feature would be useful (i.e CPE/PURLs actually exist)? mbedTLS certainly comes to mind. Thanks!

tgagneret-embedded commented:
I was thinking of this but I wasn't sure it's the best time. I'm afraid it would break some build. If we add CPE information in mbedTLS module for example and it's used with older version of Zephyr, an error will occurs when parsing the `module.yml` (because it will find unknown field).

I don't know how modules are handled, and if my use case is possible. If you think it's safe, I can update some `module.yml`.
What do you think ?

tgagneret-embedded commented:
> I was thinking of this but I wasn't sure it's the best time. I'm afraid it would break some build. If we add CPE information in mbedTLS module for example and it's used with older version of Zephyr, an error will occurs when parsing the `module.yml` (because it will find unknown field).
> 
> I don't know how modules are handled, and if my use case is possible. If you think it's safe, I can update some `module.yml`. What do you think ?

@kartben
